sys/net/network_layer/ng_sixlowpan/iphc/ng_sixlowpan_iphc.c

```diff
@@ -0,0 +1,695 @@
+}
+
+bool ng_sixlowpan_iphc_decode(ng_pktsnip_t *pkt)
+{
+    ng_netif_hdr_t *netif_hdr = pkt->next->data;
+    ng_ipv6_hdr_t *ipv6_hdr;
+    uint8_t *iphc_hdr = pkt->data;
+    uint16_t payload_offset = NG_SIXLOWPAN_IPHC_HDR_LEN;
+    ng_sixlowpan_ctx_t *ctx = NULL;
+    ng_pktsnip_t *payload;
+    ng_pktsnip_t *ipv6 = ng_pktbuf_add(NULL, NULL, sizeof(ng_ipv6_hdr_t),
+                                       NG_NETTYPE_IPV6);
+
+    if (ipv6 == NULL) {
+        DEBUG("6lo iphc: error allocating ipv6 header space\n");
+        return false;
+    }
+
+    ipv6_hdr = ipv6->data;
+
+    if (iphc_hdr[IPHC2_IDX] & NG_SIXLOWPAN_IPHC2_CID_EXT) {
+        payload_offset++;
+    }
+
+    ng_ipv6_hdr_set_version(ipv6_hdr);
+
+    switch (iphc_hdr[IPHC1_IDX] & NG_SIXLOWPAN_IPHC1_TF) {
+        case IPHC_TF_ECN_DSCP_FL:
+            ng_ipv6_hdr_set_tc(ipv6_hdr, iphc_hdr[payload_offset++]);
+            ipv6_hdr->v_tc_fl.u8[1] |= iphc_hdr[payload_offset++] & 0x0f;
+            ipv6_hdr->v_tc_fl.u8[2] |= iphc_hdr[payload_offset++];
+            ipv6_hdr->v_tc_fl.u8[3] |= iphc_hdr[payload_offset++];
+            break;
+
+        case IPHC_TF_ECN_FL:
+            ng_ipv6_hdr_set_tc_ecn(ipv6_hdr, iphc_hdr[payload_offset] >> 6);
+            ng_ipv6_hdr_set_tc_dscp(ipv6_hdr, 0);
+            ipv6_hdr->v_tc_fl.u8[1] |= iphc_hdr[payload_offset++] & 0x0f;
+            ipv6_hdr->v_tc_fl.u8[2] |= iphc_hdr[payload_offset++];
+            ipv6_hdr->v_tc_fl.u8[3] |= iphc_hdr[payload_offset++];
+            break;
+
+        case IPHC_TF_ECN_DSCP:
+            ng_ipv6_hdr_set_tc(ipv6_hdr, iphc_hdr[payload_offset++]);
+            ng_ipv6_hdr_set_fl(ipv6_hdr, 0);
+            break;
+
+        case IPHC_TF_ECN_ELIDE:
+            ng_ipv6_hdr_set_tc(ipv6_hdr, 0);
+            ng_ipv6_hdr_set_fl(ipv6_hdr, 0);
+            break;
+    }
+
+    if (!(iphc_hdr[IPHC1_IDX] & NG_SIXLOWPAN_IPHC1_NH)) {
+        ipv6_hdr->nh = iphc_hdr[payload_offset++];
+    }
+
+    switch (iphc_hdr[IPHC1_IDX] & NG_SIXLOWPAN_IPHC1_NH) {
+        case IPHC_HL_INLINE:
+            ipv6_hdr->hl = iphc_hdr[payload_offset++];
+            break;
+
+        case IPHC_HL_1:
+            ipv6_hdr->hl = 1;
+            break;
+
+        case IPHC_HL_64:
+            ipv6_hdr->hl = 64;
+            break;
+
+        case IPHC_HL_255:
+            ipv6_hdr->hl = 255;
+            break;
+    }
+
+    if (iphc_hdr[IPHC2_IDX] & NG_SIXLOWPAN_IPHC2_SAC) {
+        uint8_t sci = 0;
+
+        if (iphc_hdr[IPHC2_IDX] & NG_SIXLOWPAN_IPHC2_CID_EXT) {
+            sci = iphc_hdr[CID_EXT_IDX] >> 4;
+        }
+
+        if (iphc_hdr[IPHC2_IDX] & NG_SIXLOWPAN_IPHC2_SAM) {
+            ctx = ng_sixlowpan_ctx_lookup_id(sci);
+
+            if (ctx == NULL) {
+                DEBUG("6lo iphc: could not find source context\n");
+                return false;
+            }
+        }
+    }
+
+    switch (iphc_hdr[IPHC2_IDX] & (NG_SIXLOWPAN_IPHC2_SAC | NG_SIXLOWPAN_IPHC2_SAM)) {
```

Comment:
Isn't this the same case as in https://github.com/RIOT-OS/RIOT/pull/2806/files#diff-6c12be725eda3c6f60e9b75166d2f4b8R199?